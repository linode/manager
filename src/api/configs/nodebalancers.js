import {
  genConfig, ReducerGenerator, genActions,
  ONE, MANY, PUT, DELETE, POST,
} from '~/api/apiResultActionReducerGenerator';

export const config = genConfig({
  plural: 'nodebalancers',
  singular: 'nodebalancers',
  endpoint: id => `/nodebalancers/${id}`,
  supports: [ONE, MANY, PUT, DELETE, POST],
  subresources: {
    _configs: {
      plural: 'configs',
      singular: 'config',
      endpoint: (id, nbConfigId) => `/nodebalancers/${id}/configs/${nbConfigId}`,
      supports: [ONE, MANY, PUT, POST, DELETE],
    },
  },


});

export const actions = genActions(config);
export const { reducer } = new ReducerGenerator(config);
