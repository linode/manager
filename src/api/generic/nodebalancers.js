import {
  genConfig, ReducerGenerator, genActions,
  ONE, MANY, PUT, DELETE, POST,
} from '~/api/internal';

export const config = genConfig({
  plural: 'nodebalancers',
  endpoint: id => `/nodebalancers/${id}`,
  supports: [ONE, MANY, PUT, DELETE, POST],
  subresources: {
    _generic: {
      plural: 'generic',
      endpoint: (id, nbConfigId) => `/nodebalancers/${id}/generic/${nbConfigId}`,
      supports: [ONE, MANY, PUT, POST, DELETE],
      subresources: {
        _nodes: {
          plural: 'nodes',
          endpoint: (id, nbConfigId, nodeId) => {
            return `/nodebalancers/${id}/generic/${nbConfigId}/nodes/${nodeId}`;
          },
          supports: [ONE, MANY, PUT, POST, DELETE],
        },
      },
    },
  },


});

export const actions = genActions(config);
export const { reducer } = new ReducerGenerator(config);
