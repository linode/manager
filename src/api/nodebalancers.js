import { thunkFetch } from './apiActionReducerGenerator';

export function updateConfigSSL(data, nodebalancerId, configId) {
  return thunkFetch.post(`/nodebalancers/${nodebalancerId}/configs/${configId}/ssl`, data);
}
