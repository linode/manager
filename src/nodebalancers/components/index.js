export { default as AddNodeBalancer } from './AddNodeBalancer';
