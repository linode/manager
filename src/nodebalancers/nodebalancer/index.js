import React from 'react';
import { Route, IndexRoute } from 'react-router';

import IndexPage from './layouts/IndexPage';
import DashboardPage from './layouts/DashboardPage';
import EditConfigPage from './layouts/EditConfigPage';
import AddConfigPage from './layouts/AddConfigPage';
import SettingsPage from './layouts/SettingsPage';

export default (
  <Route path=":nbLabel" component={IndexPage}>
    <IndexRoute component={DashboardPage} />
    <Route path="settings" component={SettingsPage} />
    <Route path="configurations/create" component={AddConfigPage} />
    <Route path="configurations/:configId" component={EditConfigPage} />
  </Route>
);
