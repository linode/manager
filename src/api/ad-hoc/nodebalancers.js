import { actions } from '../generic/nodebalancers';
import { fetch } from '../fetch';

export function nodebalancerStats(nodebalancerId) {
  return async (dispatch) => {
    const { data: _stats } = await dispatch(
      fetch.get(`/nodebalancers/${nodebalancerId}/stats`)
    );
    dispatch(actions.one({ _stats }, nodebalancerId));
  };
}
