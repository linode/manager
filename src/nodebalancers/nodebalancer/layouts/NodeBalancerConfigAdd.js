import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import { ExternalLink } from 'linode-components';
import { Card, CardHeader } from 'linode-components';

import { setSource } from '~/actions/source';
import { objectFromMapByLabel } from '~/api/util';

import { ConfigForm } from '../components';


export class NodeBalancerConfigAdd extends Component {
  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));
  }

  render() {
    const { dispatch, nodebalancer } = this.props;

    return (
      <Card header={<CardHeader title="Add Config" />}>
        <p>
          {/* eslint-disable max-len */}
          Configure how your NodeBalancer listens for incoming traffic
          and connects to backend nodes. <ExternalLink to="https://www.linode.com/docs/platform/nodebalancer/getting-started-with-nodebalancers#configuring-a-nodebalancer">Learn more.</ExternalLink>
          {/* eslint-enable max-len */}
        </p>
        <ConfigForm
          dispatch={dispatch}
          nodebalancer={nodebalancer}
          submitText="Add Config"
          submitDisabledText="Adding Config"
        />
      </Card>
    );
  }
}

NodeBalancerConfigAdd.propTypes = {
  dispatch: PropTypes.func,
  nodebalancer: PropTypes.object,
};

function mapStateToProps(state, ownProps) {
  const params = ownProps.match.params;
  const nbLabel = params.nbLabel;

  const nodebalancer = objectFromMapByLabel(state.api.nodebalancers.nodebalancers, nbLabel);

  return { nodebalancer };
}

export default connect(mapStateToProps)(NodeBalancerConfigAdd);

