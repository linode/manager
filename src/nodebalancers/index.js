import React from 'react';
import { Switch, Route, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import NodeBalancersList from './layouts/NodeBalancersList';
import NodeBalancerIndex from './nodebalancer';
import Configs from './nodebalancer/configs';
import NodeBalancerConfigAdd from './nodebalancer/layouts/NodeBalancerConfigAdd';

const NodeBalancersIndex = (props) => {
  const { match: { path } } = props;
  return (
    <Switch>
      <Route component={NodeBalancerConfigAdd} path={`${path}/:nbLabel/configs/create`} />
      <Route component={Configs} path={`${path}/:nbLabel/configs/:configId`} />
      <Route component={NodeBalancerIndex} path={`${path}/:nbLabel`} />
      <Route component={NodeBalancersList} exact path={`${path}/`} />
      <Redirect to="/not-found" />
    </Switch >
  );
};

NodeBalancersIndex.propTypes = {
  match: PropTypes.shape({
    path: PropTypes.string.isRequired,
  }).isRequired,
};

export default NodeBalancersIndex;
