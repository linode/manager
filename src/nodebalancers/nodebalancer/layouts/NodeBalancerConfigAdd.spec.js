import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';
import { push } from 'react-router-redux';

import { NodeBalancerConfigAdd } from './NodeBalancerConfigAdd';

import {
  createSimulatedEvent,
  expectDispatchOrStoreErrors,
  expectRequest,
} from '~/test.helpers';
import { configsNodeBalancer } from '~/data/nodebalancers';


describe('nodebalancers/nodebalancer/layouts/AddConfigPage', () => {
  const sandbox = sinon.sandbox.create();
  const dispatch = sandbox.stub();

  afterEach(() => {
    dispatch.reset();
    sandbox.restore();
  });

  it('commits changes to the API', async () => {
    const page = await mount(
      <NodeBalancerConfigAdd
        nodebalancer={configsNodeBalancer}
        dispatch={dispatch}
      />
    );

    page.find('input[name="port"]')
      .simulate('change', createSimulatedEvent('port', 82));
    page.find('select[name="protocol"]')
      .simulate('change', createSimulatedEvent('protocol', 'http'));
    page.find('select[name="algorithm"]')
      .simulate('change', createSimulatedEvent('algorithm', 'roundrobin'));
    page.find('select[name="stickiness"]')
      .simulate('change', createSimulatedEvent('stickiness', 'none'));
    page.find('select[name="check"]')
      .simulate('change', createSimulatedEvent('check', 'none'));
    page.find('input[name="checkPassive"]')
      .simulate('change', createSimulatedEvent('checkPassive', true));
    page.find('input[name="checkInterval"]')
      .simulate('change', createSimulatedEvent('checkInterval', 0));
    page.find('input[name="checkTimeout"]')
      .simulate('change', createSimulatedEvent('checkTimeout', 30));
    page.find('input[name="checkAttempts"]')
      .simulate('change', createSimulatedEvent('checkAttempts', 3));

    dispatch.reset();
    await page.find('Form').props().onSubmit();
    expect(dispatch.callCount).toBe(1);
    await expectDispatchOrStoreErrors(dispatch.firstCall.args[0], [
      ([fn]) => expectRequest(fn, `/nodebalancers/${configsNodeBalancer.id}/configs/`, {
        method: 'POST',
        body: {
          check_timeout: 30,
          check_attempts: 3,
          check_interval: 0,
          protocol: 'http',
          algorithm: 'roundrobin',
          stickiness: 'none',
          check_passive: true,
          port: 82,
          check: 'none',
          check_path: '',
          check_body: '',
        },
      }),
      ([pushResult]) => expect(pushResult)
        .toEqual(push(`/nodebalancers/${configsNodeBalancer.label}/configs/5`)),
    ], 2, [{ id: 5 }]);
  });

  it('commits changes to the API with HTTPS', async () => {
    const page = await mount(
      <NodeBalancerConfigAdd
        nodebalancer={configsNodeBalancer}
        dispatch={dispatch}
      />
    );

    page.find('input[name="port"]')
      .simulate('change', createSimulatedEvent('port', 82));
    page.find('select[name="protocol"]')
      .simulate('change', createSimulatedEvent('protocol', 'https'));
    page.find('select[name="algorithm"]')
      .simulate('change', createSimulatedEvent('algorithm', 'roundrobin'));
    page.find('select[name="stickiness"]')
      .simulate('change', createSimulatedEvent('stickiness', 'none'));
    page.find('select[name="check"]')
      .simulate('change', createSimulatedEvent('check', 'none'));
    page.find('input[name="checkPassive"]')
      .simulate('change', createSimulatedEvent('checkPassive', true));
    page.find('input[name="checkInterval"]')
      .simulate('change', createSimulatedEvent('checkInterval', 0));
    page.find('input[name="checkTimeout"]')
      .simulate('change', createSimulatedEvent('checkTimeout', 30));
    page.find('input[name="checkAttempts"]')
      .simulate('change', createSimulatedEvent('checkAttempts', 3));
    page.find('textarea[name="sslCert"]')
      .simulate('change', createSimulatedEvent('sslCert', 'Some ssl cert'));
    page.find('textarea[name="sslKey"]')
      .simulate('change', createSimulatedEvent('sslKey', 'Some ssl key'));

    dispatch.reset();
    await page.find('Form').props().onSubmit();
    expect(dispatch.callCount).toBe(1);
    await expectDispatchOrStoreErrors(dispatch.firstCall.args[0], [
      ([fn]) => expectRequest(fn, `/nodebalancers/${configsNodeBalancer.id}/configs/`, {
        method: 'POST',
        body: {
          check_timeout: 30,
          check_attempts: 3,
          check_interval: 0,
          check_path: '',
          check_body: '',
          protocol: 'https',
          algorithm: 'roundrobin',
          stickiness: 'none',
          check_passive: true,
          port: 82,
          check: 'none',
          ssl_cert: 'Some ssl cert',
          ssl_key: 'Some ssl key',
        },
      }),
      ([pushResult]) => expect(pushResult)
        .toEqual(push(`/nodebalancers/${configsNodeBalancer.label}/configs/5`)),
    ], 2, [{ id: 5 }]);
  });
});
