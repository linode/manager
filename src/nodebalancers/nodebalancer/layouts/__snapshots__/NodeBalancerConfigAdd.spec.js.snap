// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodebalancers/nodebalancer/layouts/AddConfigPage should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NodeBalancerConfigAdd
    dispatch={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "source": "test",
              "type": "@@source/SET_SOURCE",
            },
          ],
        ],
      }
    }
    nodebalancer={
      Object {
        "_configs": Object {
          "configs": Object {
            "1": Object {
              "_nodes": Object {
                "nodes": Object {
                  "1": Object {
                    "address": "192.168.4.5:80",
                    "config_id": 0,
                    "id": 1,
                    "label": "greatest_node_ever",
                    "mode": "accept",
                    "nodebalancer_id": 23,
                    "status": "online",
                    "weight": 40,
                  },
                },
              },
              "algorithm": "roundrobin",
              "check": "none",
              "check_attempts": 3,
              "check_body": "",
              "check_interval": 0,
              "check_passive": true,
              "check_path": "",
              "check_timeout": 30,
              "cipher_suite": "recommended",
              "id": 1,
              "nodebalancer_id": 23,
              "nodes_status": Object {
                "down": 0,
                "up": 1,
              },
              "port": 80,
              "protocol": "http",
              "stickiness": "none",
            },
            "2": Object {
              "algorithm": "roundrobin",
              "check": "none",
              "check_attempts": 3,
              "check_body": "",
              "check_interval": 0,
              "check_passive": true,
              "check_path": "",
              "check_timeout": 30,
              "cipher_suite": "recommended",
              "id": 2,
              "nodebalancer_id": 23,
              "nodes_status": Object {
                "down": 0,
                "up": 0,
              },
              "port": 81,
              "protocol": "http",
              "stickiness": "none",
            },
          },
        },
        "client_conn_throttle": 0,
        "created": "2017-03-03T18:08:44",
        "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
        "id": 23,
        "ipv4": "1.1.1.1",
        "label": "nodebalancer-1",
        "region": "us-east-1a",
        "status": "new_active",
        "updated": "2017-03-06T22:00:48",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <p>
          Configure how your NodeBalancer listens for incoming traffic and connects to backend nodes. 
          <ExternalLink
            to="https://www.linode.com/docs/platform/nodebalancer/getting-started-with-nodebalancers#configuring-a-nodebalancer"
          >
            Learn more.
          </ExternalLink>
        </p>,
        <ConfigForm
          config={
            Object {
              "algorithm": "roundrobin",
              "check": "connection",
              "check_body": "",
              "check_passive": true,
              "check_path": "",
              "protocol": "http",
              "stickiness": "table",
            }
          }
          dispatch={
            [MockFunction] {
              "calls": Array [
                Array [
                  Object {
                    "source": "test",
                    "type": "@@source/SET_SOURCE",
                  },
                ],
              ],
            }
          }
          nodebalancer={
            Object {
              "_configs": Object {
                "configs": Object {
                  "1": Object {
                    "_nodes": Object {
                      "nodes": Object {
                        "1": Object {
                          "address": "192.168.4.5:80",
                          "config_id": 0,
                          "id": 1,
                          "label": "greatest_node_ever",
                          "mode": "accept",
                          "nodebalancer_id": 23,
                          "status": "online",
                          "weight": 40,
                        },
                      },
                    },
                    "algorithm": "roundrobin",
                    "check": "none",
                    "check_attempts": 3,
                    "check_body": "",
                    "check_interval": 0,
                    "check_passive": true,
                    "check_path": "",
                    "check_timeout": 30,
                    "cipher_suite": "recommended",
                    "id": 1,
                    "nodebalancer_id": 23,
                    "nodes_status": Object {
                      "down": 0,
                      "up": 1,
                    },
                    "port": 80,
                    "protocol": "http",
                    "stickiness": "none",
                  },
                  "2": Object {
                    "algorithm": "roundrobin",
                    "check": "none",
                    "check_attempts": 3,
                    "check_body": "",
                    "check_interval": 0,
                    "check_passive": true,
                    "check_path": "",
                    "check_timeout": 30,
                    "cipher_suite": "recommended",
                    "id": 2,
                    "nodebalancer_id": 23,
                    "nodes_status": Object {
                      "down": 0,
                      "up": 0,
                    },
                    "port": 81,
                    "protocol": "http",
                    "stickiness": "none",
                  },
                },
              },
              "client_conn_throttle": 0,
              "created": "2017-03-03T18:08:44",
              "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
              "id": 23,
              "ipv4": "1.1.1.1",
              "label": "nodebalancer-1",
              "region": "us-east-1a",
              "status": "new_active",
              "updated": "2017-03-06T22:00:48",
            }
          }
          submitDisabledText="Adding Config"
          submitText="Add Config"
        />,
      ],
      "className": "",
      "header": <CardHeader
        title="Add Config"
      />,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Configure how your NodeBalancer listens for incoming traffic and connects to backend nodes. ",
            <ExternalLink
              to="https://www.linode.com/docs/platform/nodebalancer/getting-started-with-nodebalancers#configuring-a-nodebalancer"
            >
              Learn more.
            </ExternalLink>,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Configure how your NodeBalancer listens for incoming traffic and connects to backend nodes. ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Learn more.",
              "to": "https://www.linode.com/docs/platform/nodebalancer/getting-started-with-nodebalancers#configuring-a-nodebalancer",
            },
            "ref": null,
            "rendered": "Learn more.",
            "type": [Function],
          },
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "config": Object {
            "algorithm": "roundrobin",
            "check": "connection",
            "check_body": "",
            "check_passive": true,
            "check_path": "",
            "protocol": "http",
            "stickiness": "table",
          },
          "dispatch": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "source": "test",
                  "type": "@@source/SET_SOURCE",
                },
              ],
            ],
          },
          "nodebalancer": Object {
            "_configs": Object {
              "configs": Object {
                "1": Object {
                  "_nodes": Object {
                    "nodes": Object {
                      "1": Object {
                        "address": "192.168.4.5:80",
                        "config_id": 0,
                        "id": 1,
                        "label": "greatest_node_ever",
                        "mode": "accept",
                        "nodebalancer_id": 23,
                        "status": "online",
                        "weight": 40,
                      },
                    },
                  },
                  "algorithm": "roundrobin",
                  "check": "none",
                  "check_attempts": 3,
                  "check_body": "",
                  "check_interval": 0,
                  "check_passive": true,
                  "check_path": "",
                  "check_timeout": 30,
                  "cipher_suite": "recommended",
                  "id": 1,
                  "nodebalancer_id": 23,
                  "nodes_status": Object {
                    "down": 0,
                    "up": 1,
                  },
                  "port": 80,
                  "protocol": "http",
                  "stickiness": "none",
                },
                "2": Object {
                  "algorithm": "roundrobin",
                  "check": "none",
                  "check_attempts": 3,
                  "check_body": "",
                  "check_interval": 0,
                  "check_passive": true,
                  "check_path": "",
                  "check_timeout": 30,
                  "cipher_suite": "recommended",
                  "id": 2,
                  "nodebalancer_id": 23,
                  "nodes_status": Object {
                    "down": 0,
                    "up": 0,
                  },
                  "port": 81,
                  "protocol": "http",
                  "stickiness": "none",
                },
              },
            },
            "client_conn_throttle": 0,
            "created": "2017-03-03T18:08:44",
            "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
            "id": 23,
            "ipv4": "1.1.1.1",
            "label": "nodebalancer-1",
            "region": "us-east-1a",
            "status": "new_active",
            "updated": "2017-03-06T22:00:48",
          },
          "submitDisabledText": "Adding Config",
          "submitText": "Add Config",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <p>
            Configure how your NodeBalancer listens for incoming traffic and connects to backend nodes. 
            <ExternalLink
              to="https://www.linode.com/docs/platform/nodebalancer/getting-started-with-nodebalancers#configuring-a-nodebalancer"
            >
              Learn more.
            </ExternalLink>
          </p>,
          <ConfigForm
            config={
              Object {
                "algorithm": "roundrobin",
                "check": "connection",
                "check_body": "",
                "check_passive": true,
                "check_path": "",
                "protocol": "http",
                "stickiness": "table",
              }
            }
            dispatch={
              [MockFunction] {
                "calls": Array [
                  Array [
                    Object {
                      "source": "test",
                      "type": "@@source/SET_SOURCE",
                    },
                  ],
                ],
              }
            }
            nodebalancer={
              Object {
                "_configs": Object {
                  "configs": Object {
                    "1": Object {
                      "_nodes": Object {
                        "nodes": Object {
                          "1": Object {
                            "address": "192.168.4.5:80",
                            "config_id": 0,
                            "id": 1,
                            "label": "greatest_node_ever",
                            "mode": "accept",
                            "nodebalancer_id": 23,
                            "status": "online",
                            "weight": 40,
                          },
                        },
                      },
                      "algorithm": "roundrobin",
                      "check": "none",
                      "check_attempts": 3,
                      "check_body": "",
                      "check_interval": 0,
                      "check_passive": true,
                      "check_path": "",
                      "check_timeout": 30,
                      "cipher_suite": "recommended",
                      "id": 1,
                      "nodebalancer_id": 23,
                      "nodes_status": Object {
                        "down": 0,
                        "up": 1,
                      },
                      "port": 80,
                      "protocol": "http",
                      "stickiness": "none",
                    },
                    "2": Object {
                      "algorithm": "roundrobin",
                      "check": "none",
                      "check_attempts": 3,
                      "check_body": "",
                      "check_interval": 0,
                      "check_passive": true,
                      "check_path": "",
                      "check_timeout": 30,
                      "cipher_suite": "recommended",
                      "id": 2,
                      "nodebalancer_id": 23,
                      "nodes_status": Object {
                        "down": 0,
                        "up": 0,
                      },
                      "port": 81,
                      "protocol": "http",
                      "stickiness": "none",
                    },
                  },
                },
                "client_conn_throttle": 0,
                "created": "2017-03-03T18:08:44",
                "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
                "id": 23,
                "ipv4": "1.1.1.1",
                "label": "nodebalancer-1",
                "region": "us-east-1a",
                "status": "new_active",
                "updated": "2017-03-06T22:00:48",
              }
            }
            submitDisabledText="Adding Config"
            submitText="Add Config"
          />,
        ],
        "className": "",
        "header": <CardHeader
          title="Add Config"
        />,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Configure how your NodeBalancer listens for incoming traffic and connects to backend nodes. ",
              <ExternalLink
                to="https://www.linode.com/docs/platform/nodebalancer/getting-started-with-nodebalancers#configuring-a-nodebalancer"
              >
                Learn more.
              </ExternalLink>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Configure how your NodeBalancer listens for incoming traffic and connects to backend nodes. ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Learn more.",
                "to": "https://www.linode.com/docs/platform/nodebalancer/getting-started-with-nodebalancers#configuring-a-nodebalancer",
              },
              "ref": null,
              "rendered": "Learn more.",
              "type": [Function],
            },
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "config": Object {
              "algorithm": "roundrobin",
              "check": "connection",
              "check_body": "",
              "check_passive": true,
              "check_path": "",
              "protocol": "http",
              "stickiness": "table",
            },
            "dispatch": [MockFunction] {
              "calls": Array [
                Array [
                  Object {
                    "source": "test",
                    "type": "@@source/SET_SOURCE",
                  },
                ],
              ],
            },
            "nodebalancer": Object {
              "_configs": Object {
                "configs": Object {
                  "1": Object {
                    "_nodes": Object {
                      "nodes": Object {
                        "1": Object {
                          "address": "192.168.4.5:80",
                          "config_id": 0,
                          "id": 1,
                          "label": "greatest_node_ever",
                          "mode": "accept",
                          "nodebalancer_id": 23,
                          "status": "online",
                          "weight": 40,
                        },
                      },
                    },
                    "algorithm": "roundrobin",
                    "check": "none",
                    "check_attempts": 3,
                    "check_body": "",
                    "check_interval": 0,
                    "check_passive": true,
                    "check_path": "",
                    "check_timeout": 30,
                    "cipher_suite": "recommended",
                    "id": 1,
                    "nodebalancer_id": 23,
                    "nodes_status": Object {
                      "down": 0,
                      "up": 1,
                    },
                    "port": 80,
                    "protocol": "http",
                    "stickiness": "none",
                  },
                  "2": Object {
                    "algorithm": "roundrobin",
                    "check": "none",
                    "check_attempts": 3,
                    "check_body": "",
                    "check_interval": 0,
                    "check_passive": true,
                    "check_path": "",
                    "check_timeout": 30,
                    "cipher_suite": "recommended",
                    "id": 2,
                    "nodebalancer_id": 23,
                    "nodes_status": Object {
                      "down": 0,
                      "up": 0,
                    },
                    "port": 81,
                    "protocol": "http",
                    "stickiness": "none",
                  },
                },
              },
              "client_conn_throttle": 0,
              "created": "2017-03-03T18:08:44",
              "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
              "id": 23,
              "ipv4": "1.1.1.1",
              "label": "nodebalancer-1",
              "region": "us-east-1a",
              "status": "new_active",
              "updated": "2017-03-06T22:00:48",
            },
            "submitDisabledText": "Adding Config",
            "submitText": "Add Config",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
